File: BAEKJOON2024/2374_같은 수로 만들기.py
result = 0
n = int(input())
if n==1:
    print(0)
else:
    i = int(input())
    
    next_i = int(input())
    if i < next_i:
        asc = True
    else:
        asc = False
    min_num, max_num = min(i, next_i),  max(i, next_i)
    
    for _ in range(n-2):
        i = next_i
        next_i = int(input())
        if i > next_i:
            if asc:
                asc = False
                if i < max_num:
                    result += (i - min_num)
                else:
                    result += (max_num - min_num)
                min_num = i
            min_num = min(i, next_i)
        elif i < next_i:
            asc = True
            max_num = max(max_num, next_i)
    
    result += (max_num - min_num)
    print(result)

File: BAEKJOON2024/2374_같은 수로 만들기.java
import java.io.*;
import java.util.*;

public class Main {
    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        int num = Integer.parseInt(br.readLine());
        int lastNum = Integer.parseInt(br.readLine());
        long result = 0;
        if (num > 1){
            int nowNum = Integer.parseInt(br.readLine());
            boolean asc = lastNum < nowNum;
            int minNum = Math.min(lastNum,nowNum);
            int maxNum = Math.max(lastNum,nowNum);

            for (int i=2 ; i < num ; i++){
                lastNum = nowNum;
                nowNum = Integer.parseInt(br.readLine());
                if (lastNum > nowNum){
                    if (asc) {
                        result += (Math.min(lastNum, maxNum) - minNum);
                        asc = false;
                        minNum = lastNum;
                    }
                    minNum = Math.min(nowNum,lastNum);
                }
                else if (lastNum < nowNum){
                    asc = true;
                    maxNum = Math.max(maxNum, nowNum);
                }
            }
            result += (maxNum - minNum);
        }
        System.out.println(result);
    }
}
